openingd/dualopend.c

```diff
@@ -95,8 +123,1303 @@ struct state {
+	 */
+	if (amount_sat_greater(remoteconf->dust_limit,
+			       remoteconf->channel_reserve)) {
+		negotiation_failed(state, am_opener,
+				   "dust_limit_satoshis %s"
+				   " too large for channel_reserve_satoshis %s",
+				   type_to_string(tmpctx, struct amount_sat,
+						  &remoteconf->dust_limit),
+				   type_to_string(tmpctx, struct amount_sat,
+						  &remoteconf->channel_reserve));
+		return false;
+	}
+
+	return true;
+}
+
+static bool is_segwit_output(struct wally_tx_output *output,
+			     const u8 *redeemscript)
+{
+	u8 *wit_prog;
+	if (tal_bytelen(redeemscript) > 0)
+		wit_prog = cast_const(u8 *, redeemscript);
+	else
+		wit_prog = cast_const(u8 *, wally_tx_output_get_script(tmpctx, output));
+
+	return is_p2wsh(wit_prog, NULL) || is_p2wpkh(wit_prog, NULL);
+}
+
+static struct wally_psbt *
+fetch_next_moves(struct state *state, const struct wally_psbt *psbt)
+{
+	u8 *msg;
+	u16 msg_type;
+	struct wally_psbt *updated_psbt;
+
+	/* Go ask lightningd what other changes we've got */
+	msg = towire_dual_open_psbt_changed(NULL, &state->channel_id, psbt);
+
+	wire_sync_write(REQ_FD, take(msg));
+	msg = wire_sync_read(tmpctx, REQ_FD);
+
+	msg_type = fromwire_peektype(msg);
+	if (msg_type == WIRE_DUAL_OPEN_FAIL) {
+		char *err;
+		if (!fromwire_dual_open_fail(msg, msg, &err))
+			master_badmsg(msg_type, msg);
+
+		peer_failed(state->pps, &state->channel_id, "%s", err);
+	} else if (msg_type == WIRE_DUAL_OPEN_PSBT_CHANGED) {
+		struct channel_id unused;
+		if (!fromwire_dual_open_psbt_changed(state, msg, &unused, &updated_psbt)) {
+			master_badmsg(msg_type, msg);
+			return NULL;
+		}
+
+		/* Does our PSBT meet requirements? */
+		if (!check_balances(updated_psbt,
+				    state->our_role == OPENER,
+				    false, /* peers input/outputs not complete */
+				    state->feerate_per_kw_funding))
+			peer_failed(state->pps, &state->channel_id,
+				    "Peer error updating tx state. "
+				    "Local funds insufficient.");
+
+		return updated_psbt;
+	} else
+		master_badmsg(msg_type, msg);
+
+	return NULL;
+}
+
+static bool send_next(struct state *state, struct wally_psbt **psbt)
+{
+	u8 *msg;
+	bool finished = false;
+	struct wally_psbt *updated_psbt;
+	struct psbt_changeset *cs = state->changeset;
+
+	/* First we check our cached changes */
+	msg = changeset_get_next(tmpctx, &state->channel_id,
+				 &cs);
+	if (msg)
+		goto sendmsg;
+
+	/* If we don't have any changes cached, go ask Alice for
+	 * what changes they've got for us */
+	updated_psbt = fetch_next_moves(state, *psbt);
+
+	/* We should always get a updated psbt back */
+	if (!updated_psbt)
+		peer_failed(state->pps, &state->channel_id,
+			    "Unable to determine next tx update");
+
+	if (psbt_has_diff(cs, *psbt, updated_psbt,
+			  &cs->added_ins, &cs->rm_ins,
+			  &cs->added_outs, &cs->rm_outs)) {
+
+		*psbt = tal_steal(state, updated_psbt);
+		msg = changeset_get_next(tmpctx, &state->channel_id,
+					 &state->changeset);
+		assert(msg);
```

Comment:
We don't need psbt_has_diff: call it psbt_get_changeset(), have it allocate and return the changeset, and then assign it to state->changeset.  It will be empty if no changes, so:

```C
state->changeset = tal_free(state->changeset);

state->changeset = psbt_get_changeset(state, *psbt, updated_psbt);
msg = changeset_get_next(tmpctx, &state->channel_id, &state->changeset);
if (msg)
     goto sendmsg;
```

Reply:
In fact, this simplifies entire logic: this entire branch goes under `if (!msg)` after `changeset_get_next`.